PCH Title Suggestions: Update Title Suggestions panel UI

vaurdan opened the pull request:
## Description
Updates the Title Suggetions Panel UI to improve user experience. 

This PR adds a new Telemetry event, as described below:

| Event Name                                           | Description                                                      |
|------------------------------------------------------|------------------------------------------------------------------|
| `title_suggestions_pinned_toggle`        | Fires when the pinned titles panel is collapsed or uncollapsed  |

## Motivation and context
* Improve the UX of the Parse.ly Content Helper
* This PR is to be merged into #2238 

## How has this been tested?
Tested locally with all the new changes validated.

vaurdan committed:
Redesign of the initial state of the Title Suggestions component

vaurdan committed:
Redesign of the initial state of the Title Suggestions component

vaurdan committed:
Redesign of the initial state of the Title Suggestions component

vaurdan committed:
Redesign the remaining functionality of the Title Suggestions component

vaurdan committed:
Add SnackBar notification on success and error.

vaurdan committed:
Merge branch 'redesign' into update/title-suggestions-ui

# Conflicts:
#	build/content-helper/editor-sidebar.asset.php
#	build/content-helper/editor-sidebar.css
#	build/content-helper/editor-sidebar.js
#	build/content-helper/excerpt-generator.asset.php
#	build/content-helper/excerpt-generator.css
#	src/content-helper/editor-sidebar/editor-sidebar.scss

vaurdan committed:
Implement new settings and clean-up documentation

acicovic commented on src/content-helper/common/components/persona-selector/component.tsx:
```diff
@@ -11,63 +11,57 @@ import {
 import { __ } from '@wordpress/i18n';
-import { Icon, chevronRight } from '@wordpress/icons';
+import { Icon, edit } from '@wordpress/icons';
+import { MoreArrow } from '../../icons/more-arrow';
```
Let's move line 15 under internal dependencies.

vaurdan replied:
Oops! Good catch.

acicovic commented on src/content-helper/common/components/persona-selector/component.tsx:
```diff
@@ -11,63 +11,57 @@ import {
  * Internal dependencies
  */
-import { PersonIcon } from '../../icons/person-icon';
+type PersonaMetadata = {
```
Let's add a comment to this type as we've done for others.

acicovic commented on src/content-helper/common/components/persona-selector/component.tsx:
```diff
@@ -77,13 +71,13 @@ type FixedPersonaProp = keyof typeof PARSELY_PERSONAS;
+ * @param {PersonaProp} persona The persona to get the value for.
  *
- * @return {string} The label for the given persona.
+ * @return {string} The value for the given persona.
```
Comments here have been updated with `value` but the const below them is named `getLabel`. I don't know if this is correct, just flagging.

acicovic commented on src/content-helper/common/components/persona-selector/style.scss:
```diff
@@ -3,9 +3,23 @@
-	margin-bottom: to_rem(10px);
 	overflow-wrap: break-word;
 	word-break: break-word;
+	height: 36px;
```
Do we need `to_rem()` here?

vaurdan committed:
Add telemetry event

acicovic commented on src/content-helper/common/components/persona-selector/style.scss:
```diff
@@ -3,9 +3,23 @@
+	}
+
+	button {
+		height: 33px;
```
Do we need `to_rem()` here?

acicovic commented on src/content-helper/common/components/persona-selector/style.scss:
```diff
@@ -21,23 +35,26 @@
-		}
+	.components-base-control__field input {
+		display: flex;
+		height: 40px;
```
Do we need `to_rem()` here?

acicovic commented on src/content-helper/common/components/tone-selector/component.tsx:
```diff
@@ -11,58 +11,55 @@ import {
 import { __ } from '@wordpress/i18n';
-import { Icon, chevronRight, megaphone } from '@wordpress/icons';
+import { Icon, edit } from '@wordpress/icons';
+import { MoreArrow } from '../../icons/more-arrow';
+import { PARSELY_PERSONAS } from '../persona-selector';
```
Let's move these to an internal dependencies section.

acicovic commented on src/content-helper/common/components/tone-selector/component.tsx:
```diff
@@ -11,58 +11,55 @@ import {
+import { MoreArrow } from '../../icons/more-arrow';
+import { PARSELY_PERSONAS } from '../persona-selector';
+
+type ToneMetadata = {
```
Let's add a comment to this type as we've done for others.

acicovic commented on src/content-helper/common/components/tone-selector/component.tsx:
```diff
@@ -72,13 +69,13 @@ type FixedToneProp = keyof typeof PARSELY_TONES;
+ * @param {ToneProp} tone The tone to get the value for.
  *
- * @return {string} The label for the given tone.
+ * @return {string} The value for the given tone.
```
Same thing I reported in a previous file about value and label. Ignore if this was done on purpose.

acicovic commented on src/content-helper/common/components/tone-selector/style.scss:
```diff
@@ -2,9 +2,23 @@
-	margin-bottom: to_rem(10px);
 	overflow-wrap: break-word;
 	word-break: break-word;
+	height: 36px;
```
Comments about possibly needing `to_rem()` apply to this file as well.

acicovic commented on src/content-helper/common/settings/types/index.ts:
```diff
@@ -1,14 +1,20 @@
  */
 export type {
 	SidebarSettings,
+	TitleSuggestionsSettings, // Part of SidebarSettings type
+	PerformanceStatsSettings, // Part of SidebarSettings type
```
```suggestion
	TitleSuggestionsSettings, // Part of SidebarSettings type.
	PerformanceStatsSettings, // Part of SidebarSettings type.
```

acicovic commented on src/content-helper/editor-sidebar/title-suggestions/component-pinned.tsx:
```diff
@@ -0,0 +1,69 @@
+import { Panel, PanelBody } from '@wordpress/components';
+import { useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+import { Telemetry } from '../../../js/telemetry/telemetry';
```
Line 7 should go to internal dependencies.

vaurdan replied:
PHPStorm import optimizer keeps messing this out... Thanks for catching it!

acicovic replied:
No worries.

acicovic commented on src/content-helper/editor-sidebar/title-suggestions/component-pinned.tsx:
```diff
@@ -0,0 +1,69 @@
+				) ) }
+			</PanelBody>
+		</Panel>
+
```
```suggestion
```
Empty line removal

acicovic commented on src/content-helper/editor-sidebar/title-suggestions/component.tsx:
```diff
@@ -117,153 +134,157 @@ export const TitleSuggestionsPanel = (): JSX.Element => {
 
-	const saveTitleOnClickHandler = async () => {
+	/**
+	 * Handle the accepted title changing, and apply the accepted title to the post.
```
```suggestion
	 * Handles the accepted title changing, and apply the accepted title to the post.
```
I've always treated `useEffect` comments as function comments, but we can change this if you think it shouldn't be the case.

acicovic commented on src/content-helper/editor-sidebar/title-suggestions/component.tsx:
```diff
@@ -117,153 +134,157 @@ export const TitleSuggestionsPanel = (): JSX.Element => {
-		</div>
-	);
+	/**
+	 * Display a snackbar notification when an error occurs.
```
```suggestion
	 * Displays a snackbar notification when an error occurs.
```

acicovic reviewed (commented):
Thank you for this!

I've done a first quick pass and left comments about minor things. I'm planning on doing a second pass once the current feedback is addressed/discussed, and I'll also give this a test try.

vaurdan committed:
Address code review feedback

acicovic commented on src/content-helper/common/components/persona-selector/component.tsx:
```diff
@@ -11,63 +11,63 @@ import {
 /**
  * List of the available personas.
- * Each persona has a label and an emoji.
+ * Each persona has a value and an icon.
```
Is this meant to be `a label and an icon`?

acicovic replied:
Or `a label and an optional icon`.

acicovic commented on src/content-helper/common/components/tone-selector/component.tsx:
```diff
@@ -11,58 +11,63 @@ import {
 /**
  * List of the available tones.
- * Each tone has a label and an emoji.
+ * Each tone has a value and an icon.
```
Is this meant to be `a label and an icon`?

vaurdan committed:
Fix docblock incorrect description in tone and persona selector

acicovic committed:
Improve function comment

acicovic commented:
I've noticed that when migrating from `develop` to this branch, (flow that will be experienced from current PCH users) Title Suggestion settings are shown like this:

<img width="283" alt="image" src="https://github.com/Parsely/wp-parsely/assets/23142906/fff845a2-dbe5-4fd6-a36e-4f41fa645f46">

It would be better to have the experience of new installations (no settings exist), which works as expected:

<img width="283" alt="image" src="https://github.com/Parsely/wp-parsely/assets/23142906/d1570489-9019-4ccc-af2a-9294c900511e">

acicovic commented:
Some additional feedback:
- If we can, it would be nice to set the focus to the respective text control when choosing Custom Tone/Persona. Makes UX faster and easier.
- Selecting a predefined Tone/Persona empties the custom Tone/Persona value. I think we should keep the custom values, but let me know if this was done on purpose.

vaurdan committed:
Fix issue with the settings migration

vaurdan committed:
Merge remote-tracking branch 'origin/update/title-suggestions-ui' into update/title-suggestions-ui

vaurdan committed:
Change Favorite/Remove to Pin/Unpin

vaurdan committed:
Change focus to input when custom tone/persona is selected

vaurdan commented:
@acicovic thanks for your feedback! 

> I've noticed that when migrating from develop to this branch, (flow that will be experienced from current PCH users) Title Suggestion settings are shown like this:

Good catch! I have addressed this and the settings are now valid when migrating to the new version.

> If we can, it would be nice to set the focus to the respective text control when choosing Custom Tone/Persona. Makes UX faster and easier.

Nice idea! I have implemented this and the focus now changes to the input text when setting a custom tone or persona.

> Selecting a predefined Tone/Persona empties the custom Tone/Persona value. I think we should keep the custom values, but let me know if this was done on purpose.

This behaviour is the same as before the UI improvements. The way the Title Suggestions store - and the respective settings store - works is by using the tone and persona fields for native values, or custom ones. So, in order to implement this suggestion, we would need to add a new fields to the Title Suggestion store (and Settings store), to store both the custom tone and the custom persona. It's definitely doable and not too complicated, but want to confirm with you first if we should move this forward either way.

vaurdan committed:
Merge branch 'redesign' into update/title-suggestions-ui

# Conflicts:
#	build/content-helper/editor-sidebar.asset.php
#	build/content-helper/editor-sidebar.css
#	build/content-helper/editor-sidebar.js

vaurdan committed:
Add built files

acicovic commented:
> Selecting a predefined Tone/Persona empties the custom Tone/Persona value. I think we should keep the custom values, but let me know if this was done on purpose.
>> This behaviour is the same as before the UI improvements.

Strange. I thought this is the way it worked before but I've tested it and you're right. No need to do anything more in this PR.

acicovic reviewed (approved):
Very exciting work, thank you! 🙂

acicovic committed:
Remove unneeded whitespace

acicovic committed:
Merge branch 'update/title-suggestions-ui' of github.com:Parsely/wp-parsely into update/title-suggestions-ui
